File: ch4/traversal/dfs_cc.py
import sys
from enum import Enum

class flag(Enum):
  UNVISITED = -1
  VISITED = -2

AL = []
dfs_num = []

def dfs(u):
  global AL
  global dfs_num

  sys.stdout.write(' '+str(u))
  dfs_num[u] = flag.VISITED.value
  for v, w in AL[u]:
    if dfs_num[v] == flag.UNVISITED.value:
      dfs(v)


def main():
  global AL
  global dfs_num

  fp = open('dfs_cc_in.txt', 'r')

  V = int(fp.readline().strip())
  AL = [[] for _ in range(V)]
  for u in range(V):
    tkn = list(map(int, fp.readline().strip().split()))
    k = tkn[0]
    for i in range(k):
      v, w = tkn[2*i+1], tkn[2*i+2]
      AL[u].append((v, w))

  print('Standard DFS Demo (the input graph must be UNDIRECTED)')
  dfs_num = [flag.UNVISITED.value] * V
  numCC = 0
  for u in range(V):
    if dfs_num[u] == flag.UNVISITED.value:
      numCC += 1
      sys.stdout.write('CC %d:' % numCC)
      dfs(u)
      sys.stdout.write('\n')
  print('There are %d connected components' % numCC);


main()

File: ch4/traversal/dfs_cc.ml
open Scanf
open Printf

type graph = int list array (* Adjacency List representation *)

let range n = List.init n (fun x -> x)

(** Returns a list of DFS orderings of all connected components *)
let dfs (adj : graph) : int list list =
  let n = Array.length adj in
  let visited = Array.make n false in
  let rec visit order u =
    visited.(u) <- true;
    adj.(u) |> List.fold_left (fun order v ->
      if visited.(v) then order
      else visit order v
    ) (u :: order)
  in
  range n |> List.fold_left (fun components u ->
    if visited.(u) then components
    else (visit [] u |> List.rev) :: components
  ) [] |> List.rev


let () =
  let ic = Scanning.open_in "dfs_cc_in.txt" in
  let sc = object method nf = bscanf ic end in
  let n = sc#nf " %d " (fun x -> x) in
  let adj = Array.init n (fun _ ->
    sc#nf " %d " (fun degree ->
      List.init degree (fun _ -> sc#nf " %d %d " (fun v w -> v))) (* Ignoring weight *)
  ) in
  
  printf "Standard DFS Demo (the input graph must be UNDIRECTED)\n";
  let ccs = dfs adj in
  ccs |> List.iteri (fun i component ->
    printf "CC %d:" (i + 1);
    component |> List.iter (printf " %d");
    printf "\n"
  );
  ccs |> List.length |> printf "There are %d connected components\n";
